### ISSUE TITLE
Harfbuzz fails to build due to wrong Visual Studio version

### ISSUE THREAD

**Issue by redagrandrei** [maps4print/azul#203]
## Description

rustup 1.18.3 (435397f48 2019-05-22)
## Version / OS

* azul version:
azul = { git = "https://github.com/maps4print/azul" } 

## Steps to Reproduce
Trying to build

## Additional Information
Getting this error:
error: failed to run custom build command for `harfbuzz-sys v0.3.0 (https://github.com/maps4print/azul-dependencies?rev=bf2933b9aac43a7003278862772250398e4fa457#bf2933b9)`

Caused by:
  process didn't exit successfully: `D:\RustProjects\azul_test\azulapp\target\debug\build\harfbuzz-sys-94af335e9d1d05e2\build-script-build` (exit code: 101)
--- stdout
cargo:rerun-if-env-changed=HARFBUZZ_SYS_NO_PKG_CONFIG

--- stderr
thread 'main' panicked at '

couldn't determine visual studio generator
if VisualStudio is installed, however, consider running the appropriate vcvars script before building this crate
', C:\Users\user\.cargo\git\checkouts\azul-dependencies-70bb1f94316762f9\bf2933b\cmake-0.1.35\src\lib.rs:687:25
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

warning: build failed, waiting for other jobs to finish...
error: build failed

**Comment by fschutt** [comment]
Did you run `vcvars.bat` as the error message told you?

**Comment by redagrandrei** [comment]
Nope, i have no idea where to find this file. As i understand its something related to Visual Studio

**Comment by redagrandrei** [comment]
@fschutt  ok i found the vcvars.bat file launched it, but it did not helped.

**Comment by Edregol** [comment]
Seem that cmake-0.1.35 does not support Visual Studio 16.2 which is the most recent Version of Visual Studio which gets installed when you follow the link provided in the installation guide.

```
fn visual_studio_generator(&self, target: &str) -> String {
        use cc::windows_registry::{find_vs_version, VsVers};

        let base = match find_vs_version() {
            Ok(VsVers::Vs15) => "Visual Studio 15 2017",
            Ok(VsVers::Vs14) => "Visual Studio 14 2015",
            Ok(VsVers::Vs12) => "Visual Studio 12 2013",
            Ok(_) => panic!("Visual studio version detected but this crate \
                             doesn't know how to generate cmake files for it, \
                             can the `cmake` crate be updated?"),
            Err(msg) => panic!(msg),
        };
[...]
```

Thats the snippet that panics in lib.rs line 677 and following. An update to cmake-0.1.40 is propably needed.

**Comment by redagrandrei** [comment]
@Edregol  Seems that we ll have to wait. Also from what i found u cannot simply install Cmake and tell visual studio to use it. 
@fschutt how can we upgrade or downgrade cmake in this case? Maybe a patch for azul is comming soon?

**Comment by metalwhale** [comment]
It seems that [`azul-dependencies`](https://github.com/maps4print/azul/blob/master/cargo/azul/Cargo.toml#L30) refers to an older version that the dependencies fix for `harfbuzz-sys 0.3.0` has not been merged yet.
[This old rev](https://github.com/maps4print/azul-dependencies/blob/bf2933b9aac43a7003278862772250398e4fa457/harfbuzz-sys-0.3.0/Cargo.toml#L27-L28) is still including `cmake-0.1.35`, which should be `cmake-0.1.40` as mentioned at [this PR](https://github.com/maps4print/azul-dependencies/commit/00e6e0ffacb008244d196df246db48f463a59992#diff-a71ab7853f2a8e0fa51a84e5a1024287).

@fschutt can you fix this for us? Can't wait for first try with azul but have to keep waiting ^^...

**Comment by fschutt** [comment]
@metalwhale Use the branch `unvendor_dependencies`, it's fixed on there. But scrolling doesn't work on that branch, that's why it's not merged into master yet.

**Comment by mnembrini** [comment]
I have the same problem but building with the `unvendor_dependencies` branch does not work for me 

```
 Compiling azul-core v0.1.0 (https://github.com/maps4print/azul?branch=unvendor_dependencies#d4eda3ca)
   Compiling freetype v0.4.1
error[E0046]: not all trait items implemented, missing: `gen_vertex_arrays_apple`, `delete_vertex_arrays_apple`, `bind_vertex_array_apple`
   --> C:\Users\user\.cargo\git\checkouts\azul-f3b1e21a5e685cad\d4eda3c\cargo\azul-core\../../azul-core\gl.rs:133:1
    |
133 | impl Gl for VirtualGlDriver {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `gen_vertex_arrays_apple`, `delete_vertex_arrays_apple`, `bind_vertex_array_apple` in implementation
    |
    = note: `gen_vertex_arrays_apple` from trait: `fn(&Self, i32) -> std::vec::Vec<u32>`
    = note: `delete_vertex_arrays_apple` from trait: `fn(&Self, &[u32])`
    = note: `bind_vertex_array_apple` from trait: `fn(&Self, u32)`

   Compiling wr_malloc_size_of v0.0.1
   Compiling plane-split v0.13.9
   Compiling webrender_api v0.60.0
error: aborting due to previous error

For more information about this error, try `rustc --explain E0046`.
error: Could not compile `azul-core`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```


OS Name:                   Microsoft Windows 10 Pro
OS Version:                10.0.17134 N/A Build 17134

**Comment by fschutt** [comment]
@mnembrini Yes, this is what I wanted to avoid by vendoring dependencies. `gleam`, an upstream dependency, didn't adhere to semantic versioning and added functions to a public trait without bumping the version number to 0.7 - and now my code is broken because cargo automatically "upgrades" gleam `0.6.19` to `0.6.20`, even though both versions of the crate are incompatible with each other. This is why I vendored the dependencies in the first place, now I'm trying to unvendor them and boom, the problem of maintainers not caring about semver is back.

**Comment by mnembrini** [comment]
Thanks for that, I read up a bit on this problem and it must be pretty frustrating. 
I added `gleam = "=0.6.19"` to my Cargo.toml and it now can compile azul-core, but I am stuck on 
azul-text-layout with this:
```
 Compiling azul-text-layout v0.1.0 (https://github.com/maps4print/azul?branch=unvendor_dependencies#d4eda3ca)
   Compiling azul-widgets v0.1.0 (https://github.com/maps4print/azul?branch=unvendor_dependencies#d4eda3ca)
error[E0308]: mismatched types
   --> C:\Users\user\.cargo\git\checkouts\azul-f3b1e21a5e685cad\d4eda3c\cargo\azul-text-layout\../../azul-text-layout\text_shaping.rs:309:73
    |
309 |         let error = FT_New_Memory_Face(ft_library, font_bytes.as_ptr(), buf_len, font_index as i64, &mut ft_face);
    |                                                                         ^^^^^^^ expected i32, found i64
help: you can convert an `i64` to `i32` and panic if the converted value wouldn't fit
    |
309 |         let error = FT_New_Memory_Face(ft_library, font_bytes.as_ptr(), buf_len.try_into().unwrap(), font_index as i64, &mut ft_face);
    |                                                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\git\checkouts\azul-f3b1e21a5e685cad\d4eda3c\cargo\azul-text-layout\../../azul-text-layout\text_shaping.rs:309:82
    |
309 |         let error = FT_New_Memory_Face(ft_library, font_bytes.as_ptr(), buf_len, font_index as i64, &mut ft_face);
    |                                                                                  ^^^^^^^^^^^^^^^^^ expected i32, found i64
help: you can convert an `i64` to `i32` and panic if the converted value wouldn't fit
    |
309 |         let error = FT_New_Memory_Face(ft_library, font_bytes.as_ptr(), buf_len, (font_index as i64).try_into().unwrap(), &mut ft_face);
    |                                                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\git\checkouts\azul-f3b1e21a5e685cad\d4eda3c\cargo\azul-text-layout\../../azul-text-layout\text_shaping.rs:318:50
    |
318 |         let error = FT_Set_Char_Size(ft_face, 0, FAKE_FONT_SIZE, DPI, DPI);
    |                                                  ^^^^^^^^^^^^^^ expected i32, found i64
help: you can convert an `i64` to `i32` and panic if the converted value wouldn't fit
    |
318 |         let error = FT_Set_Char_Size(ft_face, 0, FAKE_FONT_SIZE.try_into().unwrap(), DPI, DPI);
    |                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\git\checkouts\azul-f3b1e21a5e685cad\d4eda3c\cargo\azul-text-layout\../../azul-text-layout\text_shaping.rs:333:22
    |
333 |             x_scale: metrics.x_scale,
    |                      ^^^^^^^^^^^^^^^ expected i64, found i32

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\git\checkouts\azul-f3b1e21a5e685cad\d4eda3c\cargo\azul-text-layout\../../azul-text-layout\text_shaping.rs:334:22
    |
334 |             y_scale: metrics.y_scale,
    |                      ^^^^^^^^^^^^^^^ expected i64, found i32

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\git\checkouts\azul-f3b1e21a5e685cad\d4eda3c\cargo\azul-text-layout\../../azul-text-layout\text_shaping.rs:335:23
    |
335 |             ascender: metrics.ascender,
    |                       ^^^^^^^^^^^^^^^^ expected i64, found i32

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\git\checkouts\azul-f3b1e21a5e685cad\d4eda3c\cargo\azul-text-layout\../../azul-text-layout\text_shaping.rs:336:24
    |
336 |             descender: metrics.descender,
    |                        ^^^^^^^^^^^^^^^^^ expected i64, found i32

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\git\checkouts\azul-f3b1e21a5e685cad\d4eda3c\cargo\azul-text-layout\../../azul-text-layout\text_shaping.rs:337:21
    |
337 |             height: metrics.height,
    |                     ^^^^^^^^^^^^^^ expected i64, found i32

error[E0308]: mismatched types
   --> C:\Users\user\.cargo\git\checkouts\azul-f3b1e21a5e685cad\d4eda3c\cargo\azul-text-layout\../../azul-text-layout\text_shaping.rs:338:26
    |
338 |             max_advance: metrics.max_advance,
    |                          ^^^^^^^^^^^^^^^^^^^ expected i64, found i32

error: aborting due to 9 previous errors

For more information about this error, try `rustc --explain E0308`.
```

**Comment by fschutt** [comment]
@mnembrini yeah, I think it's because your computer is 64-bit, but using the 32-bit build of FreeType (somehow). I originally didn't watch for differences in `isize` / `i32` / `i64`, sorry. Thanks for reporting.

**Comment by tomgie** [comment]
Has this been fixed yet? I am also getting this issue.

**Comment by nintha** [comment]
There is a temp way to work around.[ link ](https://github.com/maps4print/azul/issues/163#issuecomment-538967608)

**Comment by NNemec** [comment]
For me building the current master branch with Visual Studio 2019 was easily fixed by updating to cmake-0.1.40 (pull request #236). Unless unvendor_dependencies is about to be merged within the next days, please accept that pull request to save others the struggle to sort this out.

**Comment by fschutt** [comment]
Well, this should be (temporarily) solved then? Reopen this issue if necessary.
